rllib/core/testing/tf/bc_module.py

```diff
@@ -0,0 +1,121 @@
+
+from ray.rllib.core.rl_module.rl_module import RLModule
+from ray.rllib.core.rl_module.tf.tf_rl_module import TfRLModule
+from ray.rllib.models.specs.specs_dict import ModelSpec
+from ray.rllib.models.specs.specs_tf import TFTensorSpecs
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.nested_dict import NestedDict
+
+
+class DiscreteBCTFModule(TfRLModule):
+    def __init__(
+        self,
+        input_dim: int,
+        hidden_dim: int,
+        output_dim: int,
+    ) -> None:
+        super().__init__()
+        layers = []
+
+        layers.append(tf.keras.Input(shape=(input_dim,)))
+        layers.append(tf.keras.layers.ReLU())
+        layers.append(tf.keras.layers.Dense(hidden_dim))
+        layers.append(tf.keras.layers.ReLU())
+        layers.append(tf.keras.layers.Dense(output_dim))
+
+        self.policy = tf.keras.Sequential(layers)
+        self._input_dim = input_dim
+        self._output_dim = output_dim
+
+    @override(RLModule)
+    def input_specs_exploration(self) -> ModelSpec:
+        return ModelSpec(self._default_inputs())
+
+    @override(RLModule)
+    def input_specs_inference(self) -> ModelSpec:
+        return ModelSpec(self._default_inputs())
+
+    @override(RLModule)
+    def input_specs_train(self) -> ModelSpec:
+        return ModelSpec(self._default_inputs())
+
+    @override(RLModule)
+    def output_specs_exploration(self) -> ModelSpec:
+        return ModelSpec(self._default_outputs())
+
+    @override(RLModule)
+    def output_specs_inference(self) -> ModelSpec:
+        return ModelSpec(self._default_outputs())
+
+    @override(RLModule)
+    def output_specs_train(self) -> ModelSpec:
+        return ModelSpec(self._default_outputs())
+
+    @override(RLModule)
+    def _forward_inference(self, batch: NestedDict) -> Mapping[str, Any]:
+        obs = batch[SampleBatch.OBS]
+        action_logits = self.policy(obs)
+        action_logits_inference = tf.argmax(action_logits, axis=-1)
+        action_dist = tfp.distributions.Deterministic(action_logits_inference)
+        return {"action_dist": action_dist}
+
+    @override(RLModule)
+    def _forward_exploration(self, batch: NestedDict) -> Mapping[str, Any]:
+        return self._forward_inference(batch)
+
+    @override(RLModule)
+    def _forward_train(self, batch: NestedDict) -> Mapping[str, Any]:
+        obs = batch[SampleBatch.OBS]
+        action_logits = self.policy(obs)
+        action_dist = tfp.distributions.Categorical(logits=action_logits)
+        return {"action_dist": action_dist}
+
+    @override(RLModule)
+    def get_state(self) -> Mapping[str, Any]:
+        return {"policy": self.policy.get_weights()}
+
+    @override(RLModule)
+    def set_state(self, state: Mapping[str, Any]) -> None:
+        self.policy.set_weights(state["policy"])
+
+    @override(TfRLModule)
+    def trainable_variables(self) -> Mapping[str, Any]:
```

Comment:
Can this return become `->NestedDict[tf.Tensors]` on both base class and here. So you can simply iterate through the structure leaves?

Reply:
I can but this is a flat dictionary. I don't think theres a good reason to this other than for the sake of doing it.

Reply:
don't you have to iterate over an iterable of tf.Tensors for keras optimizer? or Can you pass in an arbitrary nested dict of tf.Tensor leaves?